src/dawn/IIR/DoMethod.cpp

```diff
@@ -85,6 +107,66 @@ DoMethod::computeEnclosingAccessInterval(const int accessID, const bool mergeWit
 
 void DoMethod::setInterval(const Interval& interval) { interval_ = interval; }
 
+void DoMethod::fillWithAST(
+    const std::shared_ptr<AST> ast, const std::shared_ptr<SIR>& fullSIR,
+    iir::StencilInstantiation* instantiation,
+    const std::shared_ptr<std::vector<sir::StencilCall*>>& stackTrace,
+    const std::unordered_map<std::string, int>& localFieldnameToAccessIDMap) {
+
+  ast_->setRoot(ast->getRoot());
+
+  // Here we convert the AST of the vertical region to a flat list of statements of the stage.
+  // Further, we instantiate all referenced stencil functions.
+  DAWN_LOG(INFO) << "Inserting statements ... ";
+  StatementMapper statementMapper(fullSIR, instantiation, stackTrace, *this, interval_,
+                                  localFieldnameToAccessIDMap, nullptr);
+  ast_->accept(statementMapper);
+  DAWN_LOG(INFO) << "Inserted " << getStmts().size() << " statements";
+  if(instantiation->getOptimizerContext()->getDiagnostics().hasErrors())
+    return;
+  // Here we compute the *actual* access of each statement and associate access to the AccessIDs
+  // we set previously.
+  DAWN_LOG(INFO) << "Filling accesses ...";
+  computeAccesses(metaData_, ASTStmtToSAPMap_, sapRange());
+}
+
+DoMethod::StmtsConstIterator
+DoMethod::insertStatementAfterImpl(SAPConstIterator otherSapIt,
+                                   StmtsConstIterator& insertionPointIt) {
+  auto insertedOnlyFirstLevelIt =
```

Comment:
Isn't fixing metadata a job for the IIR tree update methods?
My understanding of what operations like insertStatementAfter should do is that they should keep consistency of the DoMethod's AST with accesses' information (all contained inside the ASTStmtToSAPMap).

Reply:
yes and no, the update method will only update the derived info objects of each of the levels of the IIR.
For example this: 
https://github.com/MeteoSwiss-APN/dawn/blob/master/src/dawn/IIR/Stage.h#L50
but not the metadata which contains global metadata information. That is built most of the times only once. 
For example stencil functions are instantiated only once 
https://github.com/MeteoSwiss-APN/dawn/blob/master/src/dawn/IIR/StencilMetaInformation.h#L350

It is most likely bad idea a map of shared_ptr to something, because precisely of the problem of cloning the object, but we have some cases done like that. 
There are not many, only a couple of shared_ptr of Stmt of the AST, but you probably need to keep an eye, since the clones will break these maps

Reply:
I understand. For now the only use case of insertStatement/s is that of moving statements from one do method to another in the split method of Stage, so we could move instead of cloning, but it doesn't seem a good idea to have it that way for other use cases we might have in the future.
I would say that insertStatement should only **add** to the maps the shared_ptr of the cloned Stmts, while I'm not sure where we should discard the orphan pairs...

Reply:
... In `PassStageSplitter`, after the call to `Stage::split`, the old stage is removed from the tree, while the new stages are inserted. 
It seems reasonable to me that the orphan pointers to AST statements/expressions that are in the Metadata should be removed by the destructor of the `DoMethod` owning them.

Reply:
I agree that we should use IDs (probably it's best to have a separate PR for that), but we wouldn't get rid of the problem of updating the Metadata when a `DoMethod` is changed/destructed/cloned, since there could be new statements/expressions and some existing ones could be removed. I'd say that `stencilFunctionInstantiations_`, `ExprToStencilFunctionInstantiationMap_`, `StencilIDToStencilCallMap_`, `fieldnameToBoundaryConditionMap_`, `BoundaryConditionToExtentsMap_` metadata fields (not sure about `stencilFunInstantiationCandidate_`) need to be updated in the `DoMethod`'s destructor, `insertStatement`/`insertStatements` and also, but I'm not sure, `clone`.

Reply:
Also, nodes of the IIR tree hold `const` references to `StencilMetaInformation`, which should be changed to non-const